compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -162,15 +171,36 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
 
         // Ensure `T: Sized` and `T: WF` obligations come last. This lets us display diagnostics
         // with more relevant type information and hide redundant E0282 errors.
-        errors.sort_by_key(|e| match e.obligation.predicate.kind().skip_binder() {
-            ty::PredicateKind::Clause(ty::ClauseKind::Trait(pred))
-                if self.tcx.is_lang_item(pred.def_id(), LangItem::Sized) =>
-            {
-                1
+        errors.sort_by_key(|e| {
+            let span = e.obligation.cause.span;
+            let outer_expn_data = span.ctxt().outer_expn_data();
+            let source_span = outer_expn_data.call_site.source_callsite();
+            let source_map = self.tcx.sess.source_map();
+
+            match e.obligation.predicate.kind().skip_binder() {
+                ty::PredicateKind::Subtype(_)
+                    if let Some(def_id) = outer_expn_data.macro_def_id
+                        && let (Some(span_file), row, col, ..) =
+                            source_map.span_to_location_info(span)
+                        && let (Some(source_file), ..) =
+                            source_map.span_to_location_info(source_span)
+                        && (self.tcx.is_diagnostic_item(sym::format_args_nl_macro, def_id)
+                            || self.tcx.is_diagnostic_item(sym::format_args_macro, def_id))
+                        && span_file.src_hash == source_file.src_hash =>
```

Comment:
Once https://github.com/rust-lang/rust/pull/142594 lands you can do away with the diagnostic item approach:

```suggestion
                    if matches!(
                        span.desugaring_kind(),
                        Some(DesugaringKind::FormatLiteral { .. })
                    ) && let (Some(_), row, col, ..) =
                        source_map.span_to_location_info(span) =>
```

Reply:
Do we not need to restrict this to `format_args_nl_macro` and `format_args_macro`? 🤔:

```rust
if matches!(
  span.desugaring_kind(),
  Some(DesugaringKind::FormatLiteral { .. })
) && (self.tcx.is_diagnostic_item(sym::format_args_nl_macro, def_id)
      || self.tcx.is_diagnostic_item(sym::format_args_macro, def_id)) =>
```